[Merged by Bors] - feat(representation_theory/monoid_algebra_basis): add some API for `k[G^n]`

101damnations opened the pull request:
---
See [this Zulip thread](https://leanprover.zulipchat.com/#narrow/stream/144837-PR-reviews/topic/.2311574.20group.20rings) for discussion. I'm not sure what to call the file...
See [this branch](https://github.com/leanprover-community/mathlib/compare/group-ring-basis) for what will be added to this file in future: a proof $k[G^{n + 1}]$ is free over $k[G]$ from the fact that $k[G^{n + 1}] \cong k[G] \otimes_k k[G^n].$
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message 
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

Commit:
add file

kim-em commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,213 @@
+/-- A term of type `Rep k G` has the natural structure of a `k[G]`-module, with an element
+`g : G` acting by `g • x = (ρ g) x` -/
+instance module' (V : Rep k G) : module (monoid_algebra k G) V :=
+representation.as_module V.ρ
```
I thought we decided this wasn't going to be an instance, and we would work directly with `\rho`?

101damnations replied:
Okay, I've been using the module instance so I can ultimately use `module.free` in the next PR, but I guess that's not strictly necessary. I've refactored to just use `\rho` & now I'm figuring out how I should change the documentation accordingly.

kim-em commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,213 @@
+
+/-- Given a `G`-action on `H`, this is `k[H]` bundled with the natural representation
+`G →* End(k[H])` as a term of type `Rep k G.` -/
+abbreviation mul_action_to_Rep : Rep k G := Rep.of (mul_action_to_ρ k G H)
```
Maybe these could be in namespaces, as `representation.of_mul_action` and `Rep.of_mul_action`?

kim-em commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,213 @@
+@[simp] lemma single_smul_single (g : G) (x : H) (r s : k) :
+  ((•) : monoid_algebra k G → mul_action_to_Rep k G H → mul_action_to_Rep k G H)
+  (finsupp.single g r) (finsupp.single x s) = finsupp.single (g • x) (r * s) :=
+by simp
```
I'm confused: why is the linter not complaining that this is a simp lemma that can be proved by simp?

antoinelab01 commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,213 @@
+variables (k G) (H : Type u) [mul_action G H]
+
+/-- A `G`-action on `H` induces a representation `G →* End(k[H])` in the natural way. -/
+@[simps] def mul_action_to_ρ : representation k G (monoid_algebra k H) :=
```
Since `H` usually doesn't have any multiplication it feels weird to use `monoid_algebra`. Why not just use `H →₀ k`?

Commit:
applying PR comments

Commit:
replace more monoid_algebra with finsupp

101damnations commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,187 @@
+
+In particular, we define morphisms of `k`-linear `G`-representations between `k[Gⁿ⁺¹]` and
+`k[G] ⊗ₖ k[Gⁿ]` (on which `G` acts by `ρ(g₁)(g₂ ⊗ x) = (g₁ * g₂) ⊗ x`) which we will later show
+define an isomorphism. (Will finish in the morning).
```
I want to talk about how this stuff will be used to show $k[G^{n + 1}]$ is free over $k[G]$, but I'm not sure how this should be stated - after passing to the Module category? I ultimately only need that it's projective anyway.

antoinelab01 replied:
Could you just state it as `module.free (monoid_algebra k G) (as_module ρ)` where `as_module` is a type synonym for the vector space which has a `module (monoid_algebra k G)` instance? (#13713 defines such `as_module`)

101damnations replied:
For sure. I was unclear on whether such type synonyms were allowed I guess - I could just add `as_module` to this PR since #13713 doesn't look like it's moving much at the moment (Edit: actually no need, I wouldn't be using it in this PR yet.)

Commit:
Merge branch 'master' of https://github.com/leanprover-community/mathlib into monoid-algebra-basis

Commit:
fix intro

ocfnash commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,189 @@
+{ to_fun := λ g, finsupp.lmap_domain k k ((•) g),
+  map_one' := by { ext x y, dsimp, simp },
+  map_mul' := λ x y, by { ext z w, simp [mul_smul] }}
+
```
I think it's worth adding:
```lean
lemma of_mul_action_apply' {G H : Type u} [group G] [mul_action G H]
  (g : G) (f : H →₀ k) (h : H) : of_mul_action k G H g f h = f (g⁻¹ • h) :=
begin
  conv_lhs { rw ← smul_inv_smul g h, },
  let h' := g⁻¹ • h,
  change of_mul_action k G H g f (g • h') = f h',
  have hg : function.injective ((•) g : H → H), { intros h₁ h₂, simp, },
  simp only [of_mul_action_apply, finsupp.lmap_domain_apply, finsupp.map_domain_apply, hg],
end
```
(it might even be giving this the `simp` attribute rather than `of_mul_action_apply` generated by `simps` since I imagine we'll often have `group G` rather than `monoid G`.)

101damnations replied:
Thank you, I have done this & also written the `simps` lemma explicitly (without a `simp` tag) as `of_mul_action_def`.

ocfnash commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,189 @@
+
+open monoid_algebra
+
+namespace representation
```
I think this section inside `namespace representation` would fit better in `representation_theory/basic.lean`.

101damnations replied:
Thanks, I've moved it - put it in a section called `mul_action`, although maybe it should be called `monoid`

ocfnash commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,189 @@
+
+end representation
+namespace Rep
+section
```
```suggestion

section monoid

```

ocfnash commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,189 @@
+abbreviation of_mul_action : Rep k G := Rep.of (representation.of_mul_action k G H)
+
+variables {k G} {n : ℕ}
+
```
```suggestion

local notation `Gⁿ` := fin n → G
local notation `Gⁿ⁺¹` := fin (n + 1) → G

```
(and if you make `n` global then you declare this once at the top of the file and reuse this in all sections)

ocfnash commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,189 @@
+/-- The `k`-linear map from `k[Gⁿ⁺¹]` to `k[G] ⊗ₖ k[Gⁿ]` sending `(g₀, ..., gₙ)`
+to `g₀ ⊗ (g₀⁻¹g₁, g₁⁻¹g₂, ..., gₙ₋₁⁻¹gₙ)` -/
+def to_tensor_aux :
+  of_mul_action k G (fin (n + 1) → G) →ₗ[k] monoid_algebra k G ⊗[k] ((fin n → G) →₀ k) :=
```
You're mixing `monoid_algebra` and `finsupp` here; is this deliberate?

I would have expected either:
```lean
  of_mul_action k G Gⁿ⁺¹ →ₗ[k] monoid_algebra k G ⊗[k] monoid_algebra k Gⁿ :=
```
or
```lean
  of_mul_action k G Gⁿ⁺¹ →ₗ[k] (G →₀ k) ⊗[k] (Gⁿ →₀ k) :=
```

More generally are there reasons for preferring `_ →₀ k` or `monoid_algebra k _` throughout this file? We should be consistent if we can.

101damnations replied:
I think since finsupp can be used everywhere, it should be; I've removed the remaining `monoid_algebra`s.

Commit:
apply review comments

Commit:
finish applying comments

Commit:
only use notation when n is implicit

ocfnash committed:
Various cosmetic tweaks.

There are no mathematical changes here. An approximate summary of
what has been done is:
  * Adjust module doc string a little
  * Adjust naming / namespacing
  * Regroup content to separate linear algebra from category-theoretic constructs
  * Regroup content to reduce implicit / explicit `variables` switching

The naming / namespacing still needs improvement.

eric-wieser commented on src/representation_theory/monoid_algebra_basis.lean:
```diff
@@ -0,0 +1,151 @@
+
+  * `representation.to_tensor`
+  * `representation.of_tensor`
+  * `Rep.of_mul_action`
```
I think these might be overindented to the point of being treated as code blocks

ocfnash replied:
Thanks, fixed.

eric-wieser commented on src/algebra/big_operators/fin.lean:
```diff
@@ -141,8 +141,36 @@ end fin
+/-- Given `f = (a₁, ..., aₙ)` in `αⁿ`, `take_prod f` is `(1, a₁, a₁a₂, ..., a₁...aₙ)` in `αⁿ⁺¹`. -/
+@[to_additive "Given `f = (a₁, ..., aₙ)` in `αⁿ`, `take_sum f` is
+`(1, a₁, a₁ + a₂, ..., a₁ + ... + aₙ)` in `αⁿ⁺¹`."]
+def take_prod (f : fin n → α) (i : fin (n + 1)) : α :=
```
It feels a bit weird to have this as a definition when we don't even have a single definition for the product of a tuple.

I wonder if something like `((list.Iio i).map f).prod` should be the preferred spelling in the longer term.
Maybe @YaelDillies has ideas.

(Using that spelling certainly is out of scope for this PR, but inlining this new def or adding a Todo comment might not be)

ocfnash replied:
This also seemed a bit weird to me and so I considered inlining. I was on the fence but persuaded myself based on the applications in this PR. I only have a moment now so I'll think again about this later today.

YaelDillies replied:
In the context of incidence algebras, this is the zeta function (yes, this is related to Riemann's zeta function!). See #11656. I'm not sure about the `fin` packaging however (#11656 gives you `(ℕ → α) → ℕ → α`, not `(fin n → α) → fin (n + 1) → α`).

Maybe @alexjbest has opinions as well.

ocfnash replied:
Thank you both for the remarks!

I thought about this some more and I'm more supportive of keeping `take_prod` as it stands (possibly adding a `TODO` if someone has a suggestion).

With regard to the `fin` packaging, I support this also because it caters for an important application. To be specific, the type `(fin n → α) → fin (n + 1) → α)` is exactly what one needs if (as in this PR) one has a (finite) `n`-tuple of elements of `α`.

Of course dissenting views are most welcome.

YaelDillies replied:
I am not in a position to judge whether we do want such a standalone definition, but I shall point out that `take_prod` is a vague name (what are we taking?). What about `cumul_prod`, `partial_prod`, `accumulate_prod` or something like this?

eric-wieser replied:
`partial_prod` or something like `tuple.scan_prod` (although I don't know if `tuple` is used anywhre) sounds good to me.

YaelDillies replied:
Note that we already have [`partial_sups`](https://leanprover-community.github.io/mathlib_docs/order/partial_sups.html#partial_sups)

ocfnash replied:
Totally agreed: `take_prod` is no good. I really like `partial_prod`. Now officially proposed.

ocfnash committed:
Reduce indentation in doc string

ocfnash committed:
Use namespace `group_cohomology.resolution` and (hopefully) improve naming

ocfnash committed:
Rename file

ocfnash committed:
Update namespacing in module doc string after 3cba155b294b524e3e4da628bd893de848610f09

ocfnash committed:
Rename `list.take_prod` to `list.partial_prod`

eric-wieser commented on src/algebra/big_operators/fin.lean:
```diff
@@ -141,8 +141,37 @@ end fin
+-/
+@[to_additive "Given `f = (a₁, ..., aₙ)` in `αⁿ`, `partial_sum f` is
+`(1, a₁, a₁ + a₂, ..., a₁ + ... + aₙ)` in `αⁿ⁺¹`."]
+def partial_prod (f : fin n → α) (i : fin (n + 1)) : α :=
```
This doesn't belong in the `list` namespace; the fact that it's implemented using lists is invisible to the user

ocfnash replied:
Good point. I've moved it to the `fin` namespace.

ocfnash committed:
Move `partial_prod` from `list` to `fin` namespace.

Also:
 * clean up legacy `to_prod` naming
 * add missing `to_additive` decorators
